keps/sig-node/00014-resource-api.md

```diff
@@ -0,0 +1,386 @@
+- Unifies with the StorageClass API. Although ResourceClass shares many similar motivations and requirements as the existing StorageClass API, they focus on different kinds of resources. StorageClass is used to represent storage resources that are stateful and contains special storage semantics. ResourceClass, on the other hand, focuses on stateless compute resources, whose usage is bound to container lifecycle and can’t be shared across multiple nodes at the same time. For these reasons, we don’t plan to unify the two APIs.
+- Resource overcommitment, fractional resource requirements, native compute resource (i.e., cpu and memory) with special metadata requirements, and group compute resources. They are out of our current scope.
+
+## Components
+### ResourceClass API
+During the initial phase, we propose to start with the following ResourceClass API spec that defines the basic ResourceClass name to the underlying node level compute resource plus metadata matching. We can extend this API to better support cluster resources and group resources in the following phases of development. However, this document will mostly focus on the design to support initial phase requirements.
+
+```golang
+// +nonNamespaced=true
+// +genclient=true
+
+type ResourceClass struct {
+        metav1.TypeMeta
+        metav1.ObjectMeta
+        Spec ResourceClassSpec
+        // +optional
+}
+
+type ResourceClassSpec struct {
+        // raw resource name. E.g.: nvidia.com/gpu
+        ResourceName string
+        // defines general resource property matching constraints.
+        // e.g.: zone in { us-west1-b, us-west1-c }; type: k80
+        MetadataRequirements metav1.LabelSelector
+        // used to compare preference of two matching ResourceClasses
+        // The purpose to introduce this field is explained more later
+        Priority int
+}
+```
+
+YAML example 1:
+```yaml
+kind: ResourceClass
+metadata:
+  name: nvidia.high.mem
+spec:
+  resourceName: "nvidia.com/nvidia-gpu"
+  labelSelector:
+    - matchExpressions:
```

Comment:
Does this still mean label selector, e.g. select nodes whose memory is `GtEq` than `15G`?

Reply:
It means select nvidia.com/gpu ComputeResource with memory GtEq than 15G. I.t., it is resource level unit, not node level. A resource can have smaller granularity than node (e.g., special CPUs with different hardware properties) or bigger granularity than node (e.g., cluster-level resource).